Std/Data/List/Basic.lean

```diff
@@ -823,6 +823,23 @@ inductive Forall₂ (R : α → β → Prop) : List α → List β → Prop
 
 attribute [simp] Forall₂.nil
 
+/--
+Decision procedure for `Forall₂ R l₁ l₂`
+-/
+def Forall₂.dec {R : α → β → Prop} [∀ a b, Decidable (R a b)] : ∀ l₁ l₂, Decidable (Forall₂ R l₁ l₂)
+  | [],    []    => isTrue .nil
+  | a::as, b::bs =>
+    if h₁ : R a b then
+      match dec as bs with
+      | isTrue h_all => isTrue (.cons h₁ h_all)
+      | isFalse h_all => isFalse fun (.cons _ h_all') => absurd h_all' h_all
```

Comment:
Verify this is tail recursive. It doesn't look like it would be, unless the compiler is abnormally smart. An alternative strategy that will be tail recursive is to call a boolean function instead and prove that the function is true iff the Forall2 statement holds.

Reply:
I think optimizing C compilers and LLVM should be able to pull that off, but I implemented it anyways